fcrepo-http-api/src/main/java/org/fcrepo/http/api/ExternalContentPathValidator.java

```diff
@@ -0,0 +1,273 @@
+ *
+ * DuraSpace licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.fcrepo.http.api;
+
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
+import static org.apache.commons.lang3.StringUtils.isEmpty;
+import static org.slf4j.LoggerFactory.getLogger;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.fcrepo.kernel.api.exception.ExternalMessageBodyException;
+import org.slf4j.Logger;
+
+/**
+ * Validates external content paths to ensure that they are within a configured allowed list of paths.
+ *
+ * @author bbpennel
+ */
+public class ExternalContentPathValidator {
+
+    private static final Logger LOGGER = getLogger(ExternalContentPathValidator.class);
+
+    private final Set<String> ALLOWED_SCHEMES = new HashSet<>(Arrays.asList("file", "http", "https"));
+
+    private final Pattern SCHEME_PATTERN = Pattern.compile("^(http|https|file):/.*");
+
+    private final Pattern RELATIVE_MOD_PATTERN = Pattern.compile(".*(^|/)\\.\\.($|/).*");
+
+    private final Pattern NORMALIZE_FILE_URI = Pattern.compile("^file:/{2,3}");
+
+    private String configPath;
+
+    private List<String> allowedList;
+
+    private boolean monitorForChanges;
+
+    private Thread monitorThread;
+
+    private boolean monitorRunning;
+
+    /**
+     * Validates that an external path is valid. The path must be an HTTP or file URI within the allow list of paths,
+     * be absolute, and contain no relative modifier.
+     *
+     * @param extPath external binary path to validate
+     * @throws ExternalMessageBodyException thrown if the path is invalid.
+     */
+    public void validate(final String extPath) throws ExternalMessageBodyException {
+        if (allowedList == null || allowedList.size() == 0) {
+            throw new ExternalMessageBodyException("External content is disallowed by the server");
```

Comment:
This functions fine, but reminds me that the constrained by file needs to be updated to be clear that it may have failed because the path/domain used was not in the 'safe' file.

Reply:
Or something to that affect.

Reply:
I'm wondering if we said something more about the allowed path.    How would a user of the system learn what the allowed path was?  That's a larger question then this PR, really.  Should we expose that information somewhere some folks using the system know what it is, or should we point them to the system administrator for more information about allowed paths?

Reply:
There could be an endpoint for retrieving the list, but my inclination at this point would be to leave it up to the administrators to expose the info in whatever way they choose to. Don't need to have fedora giving out info about how your filesystem is arranged